nx/lib/nx/random.ex

```diff
@@ -0,0 +1,374 @@
+
+  @doc """
+  Create a pseudo-random number generator (PRNG) key given an integer seed.
+
+  ## Examples
+
+      iex> Nx.Random.key(12)
+      #Nx.Tensor<
+        u32[2]
+        [0, 12]
+      >
+
+      iex> Nx.Random.key(999999999999)
+      #Nx.Tensor<
+        u32[2]
+        [232, 3567587327]
+      >
+  """
+  defn key(seed) do
+    k1 = Nx.right_shift(seed, 32)
+    k2 = Nx.bitwise_and(seed, 0xFFFFFFFF)
+
+    Nx.stack([k1, k2])
+    |> Nx.as_type({:u, 32})
+  end
+
+  @doc """
+  Splits a PRNG key into `num` new keys by adding a leading axis.
+
+    ## Examples
+
+      iex> key = Nx.Random.key(1701)
+      iex> Nx.Random.split(key, 2)
+      #Nx.Tensor<
+        u32[2][2]
+        [
+          [56197195, 1801093307],
+          [961309823, 1704866707]
+        ]
+      >
+
+      iex> key = Nx.Random.key(999999999999)
+      iex> Nx.Random.split(key, 4)
+      #Nx.Tensor<
+        u32[4][2]
+        [
+          [3959978897, 4079927650],
+          [3769699049, 3585271160],
+          [3182829676, 333122445],
+          [3185556048, 1258545461]
+        ]
+      >
+  """
+  defn split(key, num \\ 2) do
+    assert_key(key)
+
+    impl(key, Nx.iota({num, 2}))
+  end
+
+  defn fold_in(key, data) do
+    assert_key(key)
+
+    impl(key, key(data))
+  end
+
+  defn random_bits(key, opts \\ []) do
+    assert_key(key)
+    opts = keyword!(opts, shape: {1})
+
+    impl(key, Nx.iota(opts[:shape]))
+  end
+
+  defnp impl(key, count) do
+    shape = Nx.shape(count)
+
+    reshaped_key = Nx.reshape(key, {2, 1})
+
+    reshaped_count =
+      Nx.reshape(count, {:auto})
+      |> Nx.as_type({:u, 32})
+
+    threefry2x32(reshaped_key, reshaped_count)
+    |> Nx.reshape(shape)
+    |> Nx.as_type({:u, 32})
+  end
+
+  # Check count
+  defn threefry2x32(key, count) do
+    padding = Nx.size(count)
+      |> rem(2)
+
+    Nx.pad(count, 0, [{0,padding,0}])
+    |> Nx.reshape({2, :auto})
+    |> Nx.as_type({:u, 32})
+    |> threefry2x32_20(key)
+    |> Nx.reshape({:auto})
+    |> Nx.pad(0, [{0,-padding,0}])
+  end
+
+  defnp threefry2x32_20(xs, ks) do
+    rotations = Nx.tensor([[13, 15, 26, 6], [17, 29, 16, 24]], type: {:u, 8})
+    key1 = ks[0]
+    key2 = ks[1]
+
+    xs = Nx.add(ks, xs)
+
+    ks =
+      Nx.stack([
+        key2,
+        Nx.bitwise_xor(key1, key2)
+        |> Nx.bitwise_xor(0x1BD11BDA),
+        key1
+      ])
+      |> Nx.as_type({:u, 32})
+    #we lose precision on purpose
+
+    state = {xs, ks, rotations}
+
+    {_, {nxs, _, _}} =
+      while {x = 0, state}, Nx.less(x, 5) do
+        {x + 1, rolled_loop_step(x, state)}
+      end
+
+    nxs
+  end
+
+  defnp apply_round(xs, rot) do
+    y1 = Nx.sum(xs, axes: [0])
+
+    y2 =
+      rotate_left(xs[1], rot)
+      |> Nx.bitwise_xor(y1)
+
+    #losing precision on purpose due to upcasts
+    Nx.stack([y1, y2])
+    |> Nx.as_type({:u, 32})
+  end
+
+  defnp rolled_loop_step(i, {xs, ks, rs}) do
+    k1 = ks[0]
+    k2 = ks[1]
+    k3 = ks[2]
+    r1 = rs[0]
+    r2 = rs[1]

```

Comment:
I think we should receive these values separately as a 5-element tuple or as 2 separate tuples. You unwrap the tensor where, just to wrap it back up at the end of the function